lambda/src/server/slack.cljs

```diff
@@ -21,62 +26,133 @@
+
+(defn slack-user-by-email
+  "Returns slack user id for the given email address, if found"
+  [email]
+  (p/let [res (get+ "users.lookupByEmail" {:query {:email email}})]
+    ;; UNCLEAR: how does Slack know what workspace we are interested in?
+    (when (j/get res :ok)
+      ;; user contains team_id, id, is_admin
+      (j/get res :user))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; Oauth flow
+
+(j/defn req-origin [^:js {:keys [headers url query]
+                          {:keys [encrypted]} :connection}]
+  (let [host (j/get headers :host)]
+    (str "https://" host (str/replace url #"/slack.*" ""))))
+
+(j/defn install-redirect [^:js {:as req :keys [query]} res next]
+  (.redirect res
+             (str "https://slack.com/oauth/v2/authorize?"
+                  (uri/map->query-string
+                    {:scope (str/join "," scopes)
+                     :client_id (:client-id slack-config)
+                     :redirect_uri (str (req-origin req) "/slack/oauth-redirect")
+                     :state (or (:state query)
+                                (tokens/firebase-encode {:start-time (js/Date.now)
+                                                         :board-id (:board-id query)}))
+
+                     }))))
+
+(j/defn oauth-redirect [^:js {:as req :keys [body query]} res next]
+  (let [{:keys [code state]} query
+        {:keys [board-id account-id]} (tokens/firebase-decode state)]
+    (assert (and board-id account-id) "token must include board-id and account-id")
+
+    (p/let [response (post+ "oauth.v2.access"
+                            {:query {:code code
+                                     :client_id (:client-id slack-config)
+                                     :client_secret (:client-secret slack-config)
+                                     :redirect_uri (str (req-origin req) "/slack/oauth-redirect")}})]
+      (j/let [^:js {:keys [app_id
+                           bot_user_id
+                           access_token]
+                    {team-id :id team-name :name} :team
+                    {user-id :id} :authed_user} response]
+        (prn :app-id app_id :response response)
+        (p/let [user-response (get+ "users.info" {:query {:user user-id}
+                                                  :token access_token})]
+          (assert (j/get-in user-response [:user :is_admin])
+                  "Only an admin can install the Sparkboard app")
+
+          (p/all
+            [(slack-db/link-team-to-board!
+               {:slack/team-id team-id
+                :slack/bot-token access_token
+                :slack/bot-user-id bot_user_id
+                :slack/team-name team-name
+                :sparkboard/board-id board-id})
+
+             (slack-db/link-user-to-account!
+               {:slack/team-id team-id
+                :slack/user-id user-id
+                :sparkboard/account-id account-id})])
+
+          (.redirect res (str "slack://open?"
```

Comment:
Does this send the user back to Slack because they're done?

Reply:
This sends the administrator, who just added this app to a Slack workspace, to the Slack workspace. We only use the oauth-flow for this app-installation process, not for regular users.